sdk/python/lib/pulumi/_utils.py

```diff
@@ -119,3 +120,59 @@ def lazy_import(fullname):
     sys.modules[fullname] = module
     loader.exec_module(module)
     return module
+
+
+def contextproperty(_fn=None, *, default: typing.Optional[typing.Any] = None):
+    def inner(func: typing.Callable):
+        new_property = ContextProperty(
+            name=func.__qualname__, doc=func.__doc__, default=default
+        )
+        inner.__doc__ = new_property.__doc__
+        return new_property
+
+    if _fn is None:
+        return inner
+    return inner(_fn)
+
+
+class ContextProperty:
+    def __init__(
+        self, *_, name: str, doc=None, default: typing.Optional[typing.Any] = None
+    ):
+        if not doc:
+            doc = self.fget.__doc__
+        self.__doc__ = doc
+        self._name = name
+        self._data: ContextVar = ContextVar(name, default=default)
+
+    def fget(self, *_):
+        return self._data.get()
+
+    def fset(self, v: typing.Any, *_):
+        return self._data.set(v)
+
+    def fdel(self, *_):
+        return self._data.reset()
+
+    def __set__(self, _, v: typing.Any):
+        return self.fset(v)
+
+    def __get__(self, obj: typing.Optional[typing.Any], _):
+        if obj is None:
+            return self
+        return self.fget()
+
+    def __delete__(self, *_):
+        self.fdel()
+
+    def getter(self, fget):
+        prop = type(self)(fget, self.fset, self.fdel, name=self._name, doc=self.__doc__)
+        return prop
+
+    def setter(self, fset):
+        prop = type(self)(self.fget, fset, self.fdel, name=self._name, doc=self.__doc__)
+        return prop
+
+    def deleter(self, fdel):
+        prop = type(self)(self.fget, self.fset, fdel, name=self._name, doc=self.__doc__)
+        return prop
```

Comment:
This is essentially a reimplementation of the builtin `property` but wrapping values in `ContextVar`